src/utilities.jl

```diff
@@ -145,3 +145,96 @@ end
+
+### Parameters
+- `op::LBFGSOperator{T,I,F1,F2,F3}`: The L-BFGS operator `B_k`. Encodes the curvature information used to approximate the Hessian.
+- `z::AbstractVector{T}`: The vector representing `-∇f(x_k)` (negative gradient at the current iterate).
+- `σ::T`: The regularization parameter, used to shift the Hessian approximation `B_k` by adding a multiple of the identity matrix.
+- `γ_inv::T`: The inverse of the initial curvature `γ_0`, used to initialize the L-BFGS matrix.
+- `inv_Cz::AbstractVector{T}`: A preallocated vector used to store the result of the solution. It will be overwritten in the function to hold the computed `s`.
+- `p::AbstractVector{T}`: A temporary matrix used in the computation to avoid allocating new memory during the solve process.
+- `v::AbstractVector{T}`: A temporary vector for intermediate values during the solution process.
+- `u::AbstractVector{T}`: A temporary vector that stores elements of the L-BFGS vectors `a_k` or `b_k` used in the algorithm.
+
+### Returns
+- `inv_Cz::AbstractVector{T}`: The solution vector `s` such that `(B_k + σ * I) s = -∇f(x_k)`.
+
+### Method
+The function solves the system efficiently without allocating new memory, by reusing preallocated arrays `inv_Cz`, `p`, `v`, and `u`. It computes the solution iteratively, making use of the structure of the L-BFGS approximation to the Hessian.
+
+The method uses a two-loop recursion-like approach with modifications to handle the regularization term `σ`. The initial inverse Hessian approximation `B_0` is initialized using `γ_inv`.
+
+The solution is computed on the CPU, but the function can be extended for GPU compatibility.
+
+### Notes
+- `data.a[k]` and `data.b[k]` represent stored L-BFGS vectors used in the computation of `B_k`. They are loaded into `u` and updated in each iteration.
+- The computation involves vector dot products and matrix updates which are efficiently performed in place.
+
+### To-Do
+- Implement GPU support for further acceleration.
+
+### References
+@misc{erway2013shiftedlbfgssystems,
+      title={Shifted L-BFGS Systems}, 
+      author={Jennifer B. Erway and Vibhor Jain and Roummel F. Marcia},
+      year={2013},
+      eprint={1209.5141},
+      archivePrefix={arXiv},
+      primaryClass={math.NA},
+      url={https://arxiv.org/abs/1209.5141}, 
+}
+"""
+
+function solve_shifted_system!(
+  op::LBFGSOperator{T, I, F1, F2, F3},
+  z::AbstractVector{T},
+  σ::T,
+  γ_inv::T,
+  inv_Cz::AbstractVector{T},
+  p::AbstractArray{T},
+  v::AbstractVector{T},
+  u::AbstractVector{T},
+  ) where {T, I, F1, F2, F3}
+  data = op.data
+  insert = data.insert
+
+  inv_c0 = 1 / (γ_inv + σ)
+  @. inv_Cz = inv_c0 * z
+
+  max_i = 2 * data.mem
+  for i = 1:max_i
+    j = (i + 1) ÷ 2
+    k = mod(insert + j - 1, data.mem) + 1
+    u .= ((i % 2) == 0 ? data.b[k] : data.a[k])
+
+    @. p[:, i] = inv_c0 * u
+
+    for t = 1:(i - 1)
+      c0 = dot(view(p, :, t), u)
+      c1 = (-1)^(t + 1) .* v[t]
```

Comment:
Is `(-1)^(t-1)` efficient? Keeping track of whether `t` is odd or even would be more efficient.